jetty-websocket/websocket-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketServerFactory.java

```diff
@@ -303,27 +313,76 @@ public Object createWebSocket(ServletUpgradeRequest req, ServletUpgradeResponse
     @Override
     protected void doStart() throws Exception
     {
-        if(this.objectFactory == null && context != null)
+        if(this.objectFactory == null)
         {
-            this.objectFactory = (DecoratedObjectFactory) context.getAttribute(DecoratedObjectFactory.ATTR);
-            if (this.objectFactory == null)
-            {
-                throw new IllegalStateException("Unable to find required ServletContext attribute: " + DecoratedObjectFactory.ATTR);
-            }
+            this.objectFactory = findDecorator();
         }
-    
-        if(this.executor == null && context != null)
+
+        if(this.executor == null)
         {
-            ContextHandler contextHandler = ContextHandler.getContextHandler(context);
-            this.executor = contextHandler.getServer().getThreadPool();
+            this.executor = findExecutor();
         }
-        
+    
         Objects.requireNonNull(this.objectFactory, DecoratedObjectFactory.class.getName());
         Objects.requireNonNull(this.executor, Executor.class.getName());
         
         super.doStart();
     }
-    
+
+    private DecoratedObjectFactory findDecorator()
```

Comment:
perhaps make this protected so users could extend to add other ways to find decorators